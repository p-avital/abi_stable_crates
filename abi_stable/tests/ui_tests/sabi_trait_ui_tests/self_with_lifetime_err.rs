
use abi_stable::{
    sabi_trait::prelude::TD_Opaque,
    std_types::{RArc, RBox, Tuple1, Tuple2, Tuple3}
};


#[abi_stable::sabi_trait]
pub trait RFoo<'a, T: Copy + 'a> {
    fn get(&'a self) -> &'a T;
}

impl<'a, A: Copy + 'a> RFoo<'a, A> for Tuple1<A> {
    fn get(&'a self) -> &'a A {
        &self.0
    }
}

impl<'a, A: 'a, B: Copy + 'a> RFoo<'a, B> for Tuple2<A, B> {
    fn get(&'a self) -> &'a B {
        &self.1
    }
}

impl<'a, A: 'a, B: 'a, C: Copy + 'a> RFoo<'a, C> for Tuple3<A, B, C> {
    fn get(&'a self) -> &'a C {
        &self.2
    }
}

impl<'a, T> RFoo<'a, T> for RArc<T>
where
    T: 'a + Copy,
{
    fn get(&'a self) -> &'a T {
        &**self
    }
}


fn main() {
    let object = &RFoo_TO::from_ptr(RBox::new(RArc::new(76)), TD_Opaque);
    let tuple1_object = &RFoo_TO::from_ptr(RArc::new(Tuple1(100)), TD_Opaque);
    let tuple2_object = &RFoo_TO::from_value(Tuple2(101u32, 202_u32), TD_Opaque);
    let tuple3_object = &RFoo_TO::from_value(Tuple3(11, 22, 300_u32), TD_Opaque);

    assert_eq!(object.get(), &76);
    assert_eq!(tuple1_object.get(), &100);
    assert_eq!(tuple2_object.get(), &202);
    assert_eq!(tuple3_object.get(), &300);

    assert_eq!(RFoo::get(object), &76);
    assert_eq!(RFoo::get(tuple1_object), &100);
    assert_eq!(RFoo::get(tuple2_object), &202);
    assert_eq!(RFoo::get(tuple3_object), &300);

}