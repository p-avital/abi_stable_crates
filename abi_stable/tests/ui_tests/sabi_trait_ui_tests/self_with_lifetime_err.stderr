error[E0716]: temporary value dropped while borrowed
  --> tests/ui_tests/sabi_trait_ui_tests/self_with_lifetime_err.rs:42:19
   |
42 |     let object = &RFoo_TO::from_ptr(RBox::new(RArc::new(76)), TD_Opaque);
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                   |
   |                   creates a temporary which is freed while still in use
   |                   a temporary with access to the borrow is created here ...
...
57 | }
   | -
   | |
   | temporary value is freed at the end of this statement
   | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `RFoo_TO<'_, '_, RBox<()>, i32>`
   |
help: consider using a `let` binding to create a longer lived value
   |
42 ~     let binding = RFoo_TO::from_ptr(RBox::new(RArc::new(76)), TD_Opaque);
43 ~     let object = &binding;
   |

error[E0716]: temporary value dropped while borrowed
  --> tests/ui_tests/sabi_trait_ui_tests/self_with_lifetime_err.rs:43:26
   |
43 |     let tuple1_object = &RFoo_TO::from_ptr(RArc::new(Tuple1(100)), TD_Opaque);
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                          |
   |                          creates a temporary which is freed while still in use
   |                          a temporary with access to the borrow is created here ...
...
57 | }
   | -
   | |
   | temporary value is freed at the end of this statement
   | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `RFoo_TO<'_, '_, RArc<()>, i32>`
   |
help: consider using a `let` binding to create a longer lived value
   |
43 ~     let binding = RFoo_TO::from_ptr(RArc::new(Tuple1(100)), TD_Opaque);
44 ~     let tuple1_object = &binding;
   |

error[E0716]: temporary value dropped while borrowed
  --> tests/ui_tests/sabi_trait_ui_tests/self_with_lifetime_err.rs:44:26
   |
44 |     let tuple2_object = &RFoo_TO::from_value(Tuple2(101u32, 202_u32), TD_Opaque);
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                          |
   |                          creates a temporary which is freed while still in use
   |                          a temporary with access to the borrow is created here ...
...
57 | }
   | -
   | |
   | temporary value is freed at the end of this statement
   | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `RFoo_TO<'_, '_, RBox<()>, u32>`
   |
help: consider using a `let` binding to create a longer lived value
   |
44 ~     let binding = RFoo_TO::from_value(Tuple2(101u32, 202_u32), TD_Opaque);
45 ~     let tuple2_object = &binding;
   |

error[E0716]: temporary value dropped while borrowed
  --> tests/ui_tests/sabi_trait_ui_tests/self_with_lifetime_err.rs:45:26
   |
45 |     let tuple3_object = &RFoo_TO::from_value(Tuple3(11, 22, 300_u32), TD_Opaque);
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                          |
   |                          creates a temporary which is freed while still in use
   |                          a temporary with access to the borrow is created here ...
...
57 | }
   | -
   | |
   | temporary value is freed at the end of this statement
   | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `RFoo_TO<'_, '_, RBox<()>, u32>`
   |
help: consider using a `let` binding to create a longer lived value
   |
45 ~     let binding = RFoo_TO::from_value(Tuple3(11, 22, 300_u32), TD_Opaque);
46 ~     let tuple3_object = &binding;
   |
